kernel/src/net/socket/ip/stream/connecting.rs

```diff
@@ -80,35 +77,39 @@ impl ConnectingStream {
         pollee.reset_events();
     }
 
-    /// Returns `true` when `conn_result` becomes ready, which indicates that the caller should
-    /// invoke the `into_result()` method as soon as possible.
-    ///
-    /// Since `into_result()` needs to be called only once, this method will return `true`
-    /// _exactly_ once. The caller is responsible for not missing this event.
-    #[must_use]
-    pub(super) fn update_io_events(&self) -> bool {
-        if self.conn_result.read().is_some() {
-            return false;
+    pub(super) fn update_io_events(&self, pollee: &Pollee) {
+        if self.conn_result.lock().is_some() {
+            return;
         }
 
         self.bound_socket.raw_with(|socket| {
-            let mut result = self.conn_result.write();
+            let mut result = self.conn_result.lock();
             if result.is_some() {
-                return false;
+                return;
             }
 
             // Connected
             if socket.can_send() {
                 *result = Some(ConnResult::Connected);
-                return true;
+                pollee.add_events(IoEvents::OUT);
+                return;
             }
             // Connecting
             if socket.is_open() {
-                return false;
+                return;
             }
             // Refused
             *result = Some(ConnResult::Refused);
-            true
+            pollee.add_events(IoEvents::OUT);
+
+            // Add `IoEvents::OUT` because the man pages say "EINPROGRESS [..] It is possible to
+            // select(2) or poll(2) for completion by selecting the socket for writing". For
+            // details, see <https://man7.org/linux/man-pages/man2/connect.2.html>.
+            //
+            // TODO: It is better to do the state transition and let `ConnectedStream` or
+            // `InitStream` set the correct I/O events. However, the state transition is delayed
+            // because we're probably in IRQ handlers. Maybe mark the `pollee` as obsolete and
+            // re-calculate the I/O events in `poll`.
```

Comment:
Yes, but there are two occurrences of `polee.add_events(IoEvents::OUT)`. I just don't know which one to put the comments before.

Reply:
I used to think that these comments were meant to explain `pollee.add_events(IoEvents::OUT)` on line 103. Line 94 is quite intuitive; by definition, we should be adding OUT events for sockets that can send data, so additional explanation seems unnecessary.

If the comments are intended for both line 103 and line 94, it might be helpful to include an introductory line to clarify this, such as:

```rust
// The following comment applies to both line 94 and line 103:
```

Reply:
> The following comment applies to both line 94 and line 103:

Having line numbers here won't work, they are too easy to break (e.g. inserting a line in the same file but not around the `update_io_events`).

> If the comments are intended for both line 103 and line 94, it might be helpful to include an introductory line to clarify this

Well, I think the wording in the man pages is clear enough, it is not specific to a successful `connect()` or a failed `connect()`. Moreover, I additionally wrote "let **`ConnectedStream` or `InitStream`** set the correct I/O events".

Reply:
Thanks for your explanation!